File: Semestres/Estructuras de Datos/Arboles/AVL.py
class Node:
    def __init__(self,data):
        self.data = data
        self.left = None
        self.right = None
        self.parent = None
        self.balance = 0
        self.height = 1
    def __str__(self):
        s = str(self.data) 
        s+= "----> [h:"+str(self.height)
        s+= " , b:"+str(self.balance)+"]"
        return s
    
    def insert(self,node,root):
        if node.data <= self.data: 
            if self.left is None:  
                self.left = node
                node.parent = self
            else:
                self.left.insert(node,root)
        else:
            if self.right is None:  
                self.right = node
                node.parent = self
            else:
                self.right.insert(node,root)
        
        return self.mbalance(root)

    def rotate_left(self,node,root):
        parent = node.parent
        a = node
        b = node.right
        s2 = b.left
        
        a.right = s2
        if s2 is not None: s2.parent = a 
        
        b.left = a
        a.parent = b
        
        b.parent = parent
        
        if parent is None:#si es la raíz
            root=a.mbalance(root)
            root=b.mbalance(root)
            return b
        
        elif parent.left == a: parent.left = b #si no es la raíz
        elif parent.right == a: parent.right = b #si no es la raíz
        
        root = a.mbalance(root)
        root = b.mbalance(root)
        return root
    
    def rotate_right(self,node,root):
        parent = node.parent
        a = node
        b = node.left 
        s2 = b.right
        
        a.left = s2
        if s2 is not None: s2.parent = a
            
        b.right = a
        a.parent = b #nueva asignación de padre

        b.parent = parent
        
        if parent is None: #si es raíz
            root=a.mbalance(root)
            root=b.mbalance(root)
            return b
        
        elif parent.left == a: parent.left = b #si no es la raíz
        elif parent.right == a: parent.right = b  #si no es la raíz
        
        root = a.mbalance(root)
        root = b.mbalance(root)
        return root
        
    def mbalance(self,root):
        height_left = 0 if self.left is None else self.left.height
        height_right = 0 if self.right is None else self.right.height
        self.height = max(height_left+1,height_right+1)
        self.balance = height_left - height_right
        
        if self.balance<=-2 and self.right.balance<=0:
            root = self.rotate_left(self,root)
            
        elif self.balance<=-2 and self.right.balance==1:
            root = self.rotate_right(self.right,root)
            root = self.rotate_left(self,root)
            
        elif self.balance>=2 and self.left.balance==-1:
            root = self.rotate_left(self.left,root)
            root = self.rotate_right(self,root)
            
        elif self.balance>=2 and self.left.balance>=0:
            root = self.rotate_right(self,root)     
        return root
        
class AVL:
    def __init__(self):
        self.root = None
    
    def __print_node(self,node,sp):
        s = sp + node.__str__() + "\n"
        if node.left is not None: s+= self.__print_node(node.left,sp+" - ") #left
        if node.right is not None: s+= self.__print_node(node.right,sp+" - ") #right
        return s
        
    def __str__(self):
        if self.root is None: return "";
        else:  return self.__print_node(self.root,"")
        
    def insert(self,data):
        node = Node(data)
        if self.root is None:
            self.root = node
        else:
            self.root = self.root.insert(node,self.root)
            
    def __search_parent_node(self,data):
        if self.root is None:
            return None, None
        parent = None
        node = self.root
        while True:
            if data == node.data:
                return parent, node
            elif data < node.data: #izquierda
                parent = node
                node = node.left
            elif data > node.data: #derecha
                parent = node
                node = node.right
            if node is None:
                return None,None
            
    def __remove(self,parent,node):
        if node.left is None and node.right is None:
            #print('\tCaso 1')
            if parent is None: self.root = None
            elif parent.left == node:  parent.left = None
            elif parent.right == node: parent.right = None
            
        elif (node.left is None and node.right is not None) or (node.left is not None and node.right is None):
            #print('\tCaso 2')
            if parent is None:
                if node.left is not None: self.root = node.left
                elif node.right is not None: self.root = node.right
            elif parent.left == node:
                if node.left is not None: parent.left = node.left
                elif node.right is not None: parent.left = node.right
            elif parent.right == node:
                if node.left is not None: parent.right = node.left
                elif node.right is not None: parent.right = node.right 
            
        elif node.left is not None and node.right is not None:
            #print('\tCaso 3')
            parent2 = node
            node2 = node.left
            while node2.right is not None:
                parent2 = node2
                node2 = node2.right
            #print('\t-parent2:',parent2,'node2:',node2)
            self.__remove(parent2,node2)
            node2.left = node.left
            node2.right = node.right
            if parent is None: self.root = node2
            elif parent.left == node:  parent.left = node2
            elif parent.right == node: parent.right = node2
        root = node.mbalance(self.root)
        return root
        
        
    def remove(self,data):
        parent,node = self.__search_parent_node(data)
        if node is None:
            return
        #print('\t-parent:',parent,'node:',node)
        self.__remove(parent,node)
        print('\t**Delete ',node,"**\n")
        if(self.root is not None):
            self.root=self.root.mbalance(self.root)
        
    def getRoot(self):
        return self.root
            
avl = AVL()
#para right rotation
'''
avl.insert(10)
avl.insert(5)
avl.insert(1)
print(avl) '''

#para left rotation
print("SIN REMOVE:\n")
avl.insert(9)
avl.insert(10)
avl.insert(1)
avl.insert(13)
avl.insert(14)
avl.insert(15)
print(avl)

#removes
print("REMOVE:\n")
avl.remove(15)
avl.remove(14)
print(avl)

File: Semestres/Estructuras de Datos/Arboles/AVL.cpp
#include<iostream>
using namespace std;


template<class A>
class AVL{
    public:
    class Node{
        public:
        A data;
        int balance=0;
        int height=1;
        Node* left;
        Node* right;
        Node* parent;
        Node(A data){
            cout<<"\t"<<data<<endl;
            this->data = data;
            this->left = nullptr;
            this->right = nullptr;
            this->parent = nullptr;
        }
        ~Node(){
            cout<<"\t~"<<data<<endl;
        }

        string print_node(){
            string s=to_string(data);
            s+="---->[h:"+to_string(height);
            s+=" b:"+to_string(balance)+"]";
            return s;
        }

        Node* insert(Node* n,Node* root){
            if(n->data <= data){
                if(left == nullptr){
                    left = n;
                    n->parent = this; 
                }else{
                    left->insert(n,root);
                }
            }else{
                if(right == nullptr){
                    right = n;
                    n->parent = this; 
                }else{
                    right->insert(n,root);
                }  
            }
            return mbalance(root);
        }

        Node* rotate_left(Node* n,Node* root){
            Node* parent = n->parent;
            Node* a = n;
            Node* b = n->right;
            Node* s2 = b->left;

            a->right = s2;
            if(s2 != nullptr){
                s2->parent = a;
            }

            b->left = a;
            a->parent = b;

            b->parent = parent;

            if(parent == nullptr){
                root = a->mbalance(root);
                root = b->mbalance(root);
                return b;
            }
            else if(parent->left == a){
                parent->left = b;
            }
            else if(parent->right == a){
                parent->right=b;
            }

            root = a->mbalance(root);
            root = b->mbalance(root);
            return root;
        }

        Node* rotate_right(Node* n, Node* root){
            Node* parent = n->parent;
            Node* a = n;
            Node* b = n->left;
            Node* s2 = b->right;

            a->left = s2;
            if(s2 != nullptr){
                s2->parent = a;
            }

            b->right = a;
            a->parent = b;

            b->parent = parent;

            if(parent == nullptr){
                root = a->mbalance(root);
                root = b->mbalance(root);
                return b; //regresa el nodo
            }
            else if(parent->left ==a){
                parent->left = b;
            }
            else if(parent->right == a){
                parent->right = b;
            }

            root = a->mbalance(root);
            root = b->mbalance(root);
            return root; //regresa la raíz
        }

        Node* mbalance(Node* root){
            int height_left;
            int height_right;
            height_left=(left==nullptr ? 0 : left->height);
            height_right=(right==nullptr? 0 : right->height);
            height = max(height_left+1,height_right+1);
            balance = height_left - height_right;

            if(balance<=-2 && right->balance<=0){
                root = rotate_left(this,root);
            }
            else if(balance<=-2 && right->balance==1){
                root = rotate_right(right,root);
                root = rotate_left(this,root);
            }
            else if(balance>=2 && left->balance==-1){
                root = rotate_left(left,root);
                root = rotate_right(this,root);
            }
            else if(balance>=2 && left->balance>=0){
                root= rotate_right(this,root);
            }
            return root;
        }

    };
    AVL::Node* root;
    
    //Node* root;
    AVL(){
        cout<<"\tAVL()"<<endl;
        root = nullptr;
    }

    ~AVL(){
        cout<<"\t~AVL()"<<endl;
        delete_post(root);
    }

    void delete_post(Node* n){
        if(n!=nullptr){
            if(n->left) delete_post(n->left);
            if(n->right) delete_post(n->right);
            delete n;
        }
    }

    string print_avl(string space, Node* n){
        string s;
      //  s = space+to_string(n->data)+"\n";
        s = space + n->print_node()+"\n";
        if(n->left != nullptr)s+=print_avl(space+" - ",n->left);
        if(n->right != nullptr)s+=print_avl(space+" - ",n->right);
        return s;
    }

    string print_node(){
        if(root == nullptr){
            return "";
        }else{
            return print_avl("",root);
        }
    }

    void insert(A data){
        AVL::Node* n = new Node(data);
        if(root == nullptr){
            root = n;
        }else{
            root = root->insert(n,root);   
        }
    }

    Node* getRoot(){
        return root;
    }


    Node* search(A data){
          if(root == nullptr){
              return NULL; 
          }
          Node* parent = nullptr;
          Node* n = root;
          while(1){
              if(data==n->data){
                  return parent;
              }
              else if(data < n->data){ //izquierda
                  parent = n;
                  n = n->left;
              }
              else if(data > n->data){
                  parent = n;
                  n = n->right;
              }
              if(n==nullptr){
                  return NULL; 
              }
          }
      }

      void __remove(Node* parent,Node* n){
          if((n->left == nullptr) && (n->right == nullptr)){  
              cout<<"\tCaso 1"<<endl;
              if(parent == nullptr) root = NULL;
              else if(parent->left == n) parent->left = NULL;
              else if(parent->right == n)parent->right = NULL;
          } 

          else if((n->left == nullptr && n->right != nullptr) || (n->left != nullptr && n->right == nullptr)){
              cout<<"\tCaso 2"<<endl;
              if(parent == nullptr){
                  if(n->left != nullptr) root = n->left;
                  else if(n->right != nullptr) root = n->right;
              }
              else if(parent->left == n){
                  if(n->left != nullptr) parent->left = n->left;
                  else if(n->right != nullptr) parent->right = n->right;   
              }else if(parent->right == n){
                  if(n->left != nullptr) parent->left = n->left;
                  else if(n->right != nullptr)parent->right = n->right;
              }
          }
          
          else if(n->left != nullptr && n->right != nullptr){
              cout<<"\tCaso 3"<<endl;
              Node* parent2 = n;
              Node* n2 = n->left;
              while(n2->right != nullptr){
                  parent2 = n2;
                  n2 = n2->right;
              }
             // cout<<"\t-Parent 2: "<<parent2->data<<" Node2: "<<n2->data<<endl;
              __remove(parent2,n2);
              n2->left = n->left;
              n2->right = n->right;
              if(parent == nullptr) root = n2;
              else if(parent->left == n) parent->left = n2;
              else if(parent->right == n) parent->right = n2;
          }

      }

      void remove(A data){
      Node *parent = search(data);
      Node* n;
      if(parent!=nullptr) {
          if(parent -> left && parent -> left -> data == data){
              n=parent->left;
          }else{
              n=parent->right;
          }
      }
      else{
          n = root;
      }
      if(!n || n->data!=data){
          return;
      }
     // cout<<"\t-Parent: "<<parent->data<<" Node: "<<n->data<<endl;
      __remove(parent, n);
      cout<<"\t**Delete "<<n->data<<"**"<<endl;
      delete n;

      //BALANCEO DESPUÉS DE ELIMINAR
      
      
    }


};


int main(){
    AVL<int> avl;
    cout<<"\n****ARBOLITO AVL****"<<"\n";
    /*avl.insert(10);
    avl.insert(5);
    avl.insert(1);
    cout<<avl.print_node()<<endl;*/


      string str = "SaraCarolinaGomezDelgado";   //BST for char
    for(auto i:str) avl.insert(i);

    /*avl.insert(9);
    avl.insert(10);
    avl.insert(1);
    avl.insert(13);
    avl.insert(14);
    avl.insert(15);*/
    cout<<"SIN REMOVE:\n"<<avl.print_node()<<endl;

    /*avl.remove(14);
    avl.remove(15);
    cout<<"REMOVE:\n"<<avl.print_node()<<endl;*/
}